src/hypothesis/strategies.py

```diff
@@ -783,18 +785,90 @@ def seed_random(seed):
 @cacheable
 @defines_strategy
 def builds(target, *args, **kwargs):
-    """Generates values by drawing from args and kwargs and passing them to
-    target in the appropriate argument position.
+    """Generates values by drawing from ``args`` and ``kwargs`` and passing
+    them to ``target`` in the appropriate argument position.
 
-    e.g. builds(target, integers(), flag=booleans()) would draw an
-    integer i and a boolean b and call target(i, flag=b).
+    e.g. ``builds(target, integers(), flag=booleans())`` would draw an
+    integer ``i`` and a boolean ``b`` and call ``target(i, flag=b)``.
+
+    If ``target`` has type annotations, they will be used to infer a strategy
+    for required arguments that were not passed to builds.  You can also tell
+    builds to infer a strategy for an optional argument by passing the special
+    value :const:`hypothesis.strategies.infer` as a keyword argument to
+    builds, instead of a strategy for that argument to ``target``.
 
     """
+    from hypothesis.searchstrategy import types
+
+    hints = types.get_type_hints(target)
+    inferred = {kw: types.from_type(type_) for kw, type_ in hints.items()}
+
+    if infer in args:
+        raise InvalidArgument(
+            'infer may only be given as a keyword argument to builds()'
```

Comment:
Simply that it's meaningless for arguments without a default value - they have to be inferred anyway.  Definitely needs a better message though.